doc/CMakeLists.txt

```diff
@@ -68,9 +68,76 @@ set(DOC_LOG ${CMAKE_CURRENT_BINARY_DIR}/doc.log)
 set(DOXY_LOG ${CMAKE_CURRENT_BINARY_DIR}/doxy.log)
 set(SPHINX_LOG ${CMAKE_CURRENT_BINARY_DIR}/sphinx.log)
 set(DOC_WARN ${CMAKE_CURRENT_BINARY_DIR}/doc.warnings)
+set(CONTENT_OUTPUTS ${CMAKE_CURRENT_BINARY_DIR}/extracted-content.txt)
 
 configure_file(${DOXYFILE_IN} ${DOXYFILE_OUT} @ONLY)
 
+# This command is used to copy all documentation source files into the
+# build directory. Sphinx requires a single documentation root, but
+# Zephyr's documentation is scattered around the tree in samples/,
+# boards/, and doc/, so we need to copy those files into a single
+# place in the build directory.
+set(EXTRACT_CONTENT_COMMAND
+  ${CMAKE_COMMAND} -E env
+  ${PYTHON_EXECUTABLE} scripts/extract_content.py
+  # Save the outputs for processing by this list file.
+  --outputs ${CONTENT_OUTPUTS}
+  # Ignore any files in the output directory.
+  --ignore ${CMAKE_CURRENT_BINARY_DIR}
+  # Copy all files in doc to the rst folder.
+  "*:doc:${RST_OUT}"
+  # Copy the .rst files in samples/ and boards/ to the rst folder, and
+  # also the doc folder inside rst.
+  #
+  # Some files refer to items in samples/ and boards/ relative to
+  # their actual position in the Zephyr tree. For example, in
+  # subsystems/sensor.rst:
+  #
+  # .. literalinclude:: ../../samples/sensor/mcp9808/src/main.c
+  #
+  # We make an additional copy as a hackaround so these references
+  # work.
+  "*.rst:samples:${RST_OUT}" "*.rst:boards:${RST_OUT}"
+  "*.rst:samples:${RST_OUT}/doc" "*.rst:boards:${RST_OUT}/doc")
+
+# Run the content extraction command at configure time in order to
+# produce lists of input and output files, which are respectively its
+# dependencies and outputs. This also causes the content files
+# to be copied for the first time.
+execute_process(
+  COMMAND ${EXTRACT_CONTENT_COMMAND}
+  WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}
+  RESULT_VARIABLE RET)
+
+if (NOT RET EQUAL 0)
+  message(FATAL_ERROR "Cannot prepare documentation content extraction")
+endif()
+
+# Load the outputs listing from extract_content.py. This is a file
+# where each pair of lines is a source file in the Zephyr tree and a
+# destination file in the build directory.
+file(STRINGS ${CONTENT_OUTPUTS} EXTRACT_CONTENT_RAW)
+
+# Build up a list of output files for later use in target/command
+# DEPENDS. Make sure each output file gets updated if the
+# corresponding input file changes.
+set(EXTRACT_CONTENT_OUTPUTS)
+while(EXTRACT_CONTENT_RAW)
+  list(GET EXTRACT_CONTENT_RAW 0 SRC)
+  list(GET EXTRACT_CONTENT_RAW 1 DST)
+
+  add_custom_command(
+    COMMAND ${CMAKE_COMMAND} -E copy ${SRC} ${DST}
+    DEPENDS ${SRC}
+    OUTPUT ${DST}
+    )
+
+  list(REMOVE_AT EXTRACT_CONTENT_RAW 0 1)
+  list(APPEND EXTRACT_CONTENT_OUTPUTS ${DST})
+endwhile()
+
+add_custom_target(content DEPENDS ${EXTRACT_CONTENT_OUTPUTS})
```

Comment:
What if we add a new .rst file to the `doc/` folder or elsewhere? do we need then a clean build for it to be included?

Reply:
You just need to re-run cmake. The execute_process() will then run again, and the dependencies will be updated. This is basically the same thing that happens in source code that uses GLOB to find its C files.

Reply:
This should be documented in `doc/README` then I think, since CMake won't be run automatically.